### PR TITLE
[RW-600] Adding concept search (for use in notebooks / field set creation)

### PR DESCRIPTION
**Pull request by danqrodney**
(Still working on tests.)

@biopete I think at some point we may want to move the controller code and maybe even API YAML definition for this into common-api, so we can share between data browser and workbench. For now, just putting this in workbench.

### CODE REVIEW

File: api/src/main/resources/client_api.yaml [hunk 1]
```diff
@@ -74,6 +74,57 @@ paths:
           schema:
             $ref: "#/definitions/MaterializeCohortResponse"
 
+  /v1/workspaces/{workspaceNamespace}/{workspaceId}/concepts:
+      get:
+        security: []
```
Comment (calbach) [inline api/src/main/resources/client_api.yaml]: This shouldn't be used in combination with a workspace path. You should need access to a workspace in order to call any method on it, even to know whether or not it exists. Should require auth or not make this workspace-centric.

File: api/src/main/resources/client_api.yaml [hunk 2]
```diff
@@ -74,6 +74,57 @@ paths:
...
+            name: query
+            type: string
+            required: true
+            description: >
+              A query string that can be used to match a subset of the name (case-insensitively),
+              the entire code value (case-insensitively), or the concept ID.
+          - in: query
+            name: standard_concept
+            type: boolean
+            required: false
```
Comment (calbach) [inline api/src/main/resources/client_api.yaml]: I'm pretty sure false is the default, would remove these

File: api/src/main/resources/client_api.yaml [hunk 3]
```diff
@@ -74,6 +74,57 @@ paths:
...
+            description: >
+              True if only standard concepts should be returned, false if any concepts
```
Comment (calbach) [inline api/src/main/resources/client_api.yaml]: Defaults to..

File: api/src/main/resources/client_api.yaml [hunk 4]
```diff
@@ -74,6 +74,57 @@ paths:
...
+            name: vocabulary_id
+            type: string
+            required: false
+            description: The vocabulary ID for the concepts returned (e.g. SNOMED, RxNorm)
+          - in: query
+            name: domain_id
+            type: string
+            required: false
+            description: >
+             The domain ID for the concepts returned (e.g. Observation, Drug). TODO: consider
```
Comment (calbach) [inline api/src/main/resources/client_api.yaml]: Is the full list documented somewhere? Here and above

File: api/src/main/resources/client_api.yaml [hunk 5]
```diff
@@ -74,6 +74,57 @@ paths:
...
+             changing this to be an enum which is mapped to underlying domain ID value(s).
+          - in: query
```
Comment (calbach) [inline api/src/main/resources/client_api.yaml]: note: pagination is difficult to add later, as it is a breaking change. Even if the initial implementation is simplistic. Consider adding it

File: api/src/main/resources/client_api.yaml [hunk 6]
```diff
@@ -311,3 +362,58 @@ definitions:
   Operator:
     type: string
     enum: [EQUAL, NOT_EQUAL, LESS_THAN, GREATER_THAN, LESS_THAN_OR_EQUAL_TO, GREATER_THAN_OR_EQUAL_TO, LIKE, IN, BETWEEN]
+
+  Concept:
```
Comment (calbach) [inline api/src/main/resources/client_api.yaml]: Please add a description, this terminology is probably only familiar to OMOP users.

File: api/src/test/java/org/pmiops/workbench/api/ConceptsControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    result.setConceptClassId(conceptClassId);
+    result.setVocabularyId(vocabularyId);
+    result.setDomainId(domainId);
+    result.setCountValue(countValue);
+    result.setPrevalence(prevalence);
+    return result;
+  }
+
+  private static Concept makeClientConcept(org.pmiops.workbench.cdr.model.Concept dbConcept) {
+    Concept concept = new Concept();
```
Comment (calbach) [inline api/src/test/java/org/pmiops/workbench/api/ConceptsControllerTest.java]: Should really use the chained setters here:

```
return new Concept()
  .conceptId(condeptId)
  .domainId(domainId)
  ...;
```

File: api/src/test/java/org/pmiops/workbench/api/ConceptsControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+  @Test
+  public void testConceptsSearchNameOneMatch() throws Exception {
+    conceptDao.save(CONCEPT_1);
+    conceptDao.save(CONCEPT_2);
+    assertResults(
+        conceptsController.searchConcepts("a", null, null,
+            null, null), CLIENT_CONCEPT_1);
+  }
+
+  private static org.pmiops.workbench.cdr.model.Concept makeConcept(
```
Comment (calbach) [inline api/src/test/java/org/pmiops/workbench/api/ConceptsControllerTest.java]: This method signature is causing me pain. If possible, I'd say:
- define the client models inline above with the chaining approach
- switch the converter to the opposite direction
- remove this function

File: api/src/test/java/org/pmiops/workbench/test/TestDialect.java [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+import org.hibernate.dialect.MySQLDialect;
+import org.hibernate.dialect.function.SQLFunctionTemplate;
+import org.hibernate.type.StandardBasicTypes;
+
+public class TestDialect extends MySQLDialect {
+
+  public TestDialect() {
+    super();
+    // For in-memory tests, use LIKE for full text searches.
+    registerFunction("match", new SQLFunctionTemplate(StandardBasicTypes.DOUBLE,
```
Comment (calbach) [inline api/src/test/java/org/pmiops/workbench/test/TestDialect.java]: What does this have to do with a DOUBLE? Am I missing something?

File: common-api/src/main/java/org/pmiops/workbench/cdr/MySQLDialect.java [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+import org.hibernate.dialect.MySQL5Dialect;
+import org.hibernate.dialect.function.SQLFunctionTemplate;
+import org.hibernate.type.StandardBasicTypes;
+
+public class MySQLDialect extends MySQL5Dialect {
+
+  public MySQLDialect() {
+    super();
+    registerFunction("match", new SQLFunctionTemplate(StandardBasicTypes.DOUBLE,
```
Comment (calbach) [inline common-api/src/main/java/org/pmiops/workbench/cdr/MySQLDialect.java]: Please add a comment

File: common-api/src/main/java/org/pmiops/workbench/cdr/dao/ConceptService.java [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+  private final EntityManager entityManager;
+
+  public ConceptService(EntityManager entityManager) {
+    this.entityManager = entityManager;
+  }
+
+  public static final String STANDARD_CONCEPT_CODE = "S";
+
+  public Slice<Concept> searchConcepts(String query,
+      Boolean standardConcept, String vocabularyId, String domainId,
```
Comment (calbach) [inline common-api/src/main/java/org/pmiops/workbench/cdr/dao/ConceptService.java]: I would make this a `boolean` and push the default-picking logic up to the API level

File: api/src/main/java/org/pmiops/workbench/api/ConceptsController.java [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+@RestController
+public class ConceptsController implements ConceptsApiDelegate {
+
+  private static final Integer DEFAULT_MAX_RESULTS = 20;
+  private static final int MAX_MAX_RESULTS = 1000;
+
+  private final ConceptService conceptService;
+
+  private static final Function<Concept, org.pmiops.workbench.model.Concept> TO_CLIENT_CONCEPT =
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/api/ConceptsController.java]: nit: we usually import the client models, and full-qualify DB models in our other controllers

File: api/src/main/java/org/pmiops/workbench/api/ConceptsController.java [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+      (concept) ->  {
+        org.pmiops.workbench.model.Concept result = new org.pmiops.workbench.model.Concept();
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/api/ConceptsController.java]: chain (see comment in tests)

File: api/src/main/resources/client_api.yaml (hunk 1 above)
Comment (danqrodney) [reply api/src/main/resources/client_api.yaml]: Auth required now, good catch.

File: api/src/main/resources/client_api.yaml (hunk 2 above)
Comment (danqrodney) [reply api/src/main/resources/client_api.yaml]: Done.

File: api/src/main/resources/client_api.yaml (hunk 3 above)
Comment (danqrodney) [reply api/src/main/resources/client_api.yaml]: Done, changed the docs to indicate "false" means non-standard only, and default is everything.

File: api/src/main/resources/client_api.yaml (hunk 4 above)
Comment (danqrodney) [reply api/src/main/resources/client_api.yaml]: Not right now. I plan to have constants in the client library for this in the short term; long term I was going to use an enum here.

File: api/src/main/resources/client_api.yaml (hunk 5 above)
Comment (danqrodney) [reply api/src/main/resources/client_api.yaml]: OK, I'll think about this. I was of the mind that there isn't a good use case for returning more than one (largish) page of results here, and doing whatever paging you want to do client-side. And the current use case (displaying the results in a data frame in a notebook) definitely doesn't lend itself to pagination. But I could imagine adding pagination here anyway, just in case.

File: api/src/main/resources/client_api.yaml (hunk 6 above)
Comment (danqrodney) [reply api/src/main/resources/client_api.yaml]: Done.

File: api/src/test/java/org/pmiops/workbench/api/ConceptsControllerTest.java (hunk 8 above)
Comment (danqrodney) [reply api/src/test/java/org/pmiops/workbench/api/ConceptsControllerTest.java]: Done.

File: api/src/test/java/org/pmiops/workbench/api/ConceptsControllerTest.java (hunk 7 above)
Comment (danqrodney) [reply api/src/test/java/org/pmiops/workbench/api/ConceptsControllerTest.java]: Done.

File: api/src/test/java/org/pmiops/workbench/test/TestDialect.java (hunk 9 above)
Comment (danqrodney) [reply api/src/test/java/org/pmiops/workbench/test/TestDialect.java]: I'm not sure why it is this way, but it is this way... see:

https://pavelmakhov.com/2016/09/jpa-custom-function

Adding a comment.

File: common-api/src/main/java/org/pmiops/workbench/cdr/MySQLDialect.java (hunk 10 above)
Comment (danqrodney) [reply common-api/src/main/java/org/pmiops/workbench/cdr/MySQLDialect.java]: Done.

File: common-api/src/main/java/org/pmiops/workbench/cdr/dao/ConceptService.java (hunk 11 above)
Comment (danqrodney) [reply common-api/src/main/java/org/pmiops/workbench/cdr/dao/ConceptService.java]: So actually, I'm treating this as tri-state: standardConcept being null means both standard and non-standard concepts are returned, whereas true or false return only standard or non-standard concepts respectively.

If you prefer we could not bother with the "only non-standard concepts" or have an enum I guess.

File: api/src/main/java/org/pmiops/workbench/api/ConceptsController.java (hunk 12 above)
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/api/ConceptsController.java]: Done.

File: api/src/main/java/org/pmiops/workbench/api/ConceptsController.java (hunk 13 above)
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/api/ConceptsController.java]: Done.

File: common-api/src/main/java/org/pmiops/workbench/cdr/dao/ConceptService.java (hunk 11 above)
Comment (calbach) [reply common-api/src/main/java/org/pmiops/workbench/cdr/dao/ConceptService.java]: In general I don't like using null/undefined as a sentinel value for a boolean, so would prefer to collapse these or switch to an enum. On that decision, the usual API rule is: if a field has two possible values now, but could have more later, use an enum over a boolean.

File: api/src/main/java/org/pmiops/workbench/api/ConceptsController.java [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+      String workspaceId, String query,
+      Boolean standardConcept, String vocabularyId, String domainId, Integer maxResults) {
+    workspaceService.enforceWorkspaceAccessLevel(
+        workspaceNamespace, workspaceId, WorkspaceAccessLevel.READER);
+    Workspace workspace = workspaceService.getRequired(workspaceNamespace, workspaceId);
+    CdrVersion cdrVersion = workspace.getCdrVersion();
+    CdrVersionContext.setCdrVersion(cdrVersion);
+    if (maxResults == null) {
+      maxResults = DEFAULT_MAX_RESULTS;
+    } else if (maxResults < 1 || maxResults > MAX_MAX_RESULTS) {
```
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/api/ConceptsController.java]: Done.

**Comment by calbach** [comment]
Looks good, just holding out for resolution of the boolean/enum issue.

File: api/src/main/java/org/pmiops/workbench/api/ConceptsController.java [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+      throw new BadRequestException("Invalid value for maxResults: " + maxResults);
+    }
+    if (Strings.isNullOrEmpty(query.trim())) {
```
[the remark this answers is not in the document]
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/api/ConceptsController.java]: Done.

File: common-api/src/main/java/org/pmiops/workbench/cdr/dao/ConceptService.java (hunk 11 above)
Comment (danqrodney) [reply common-api/src/main/java/org/pmiops/workbench/cdr/dao/ConceptService.java]: OK. I changed this to be an enum, as well as domain (which is now being mapped from a set of domain enum values to underlying domain IDs in ConceptsController.)

Note that there's some duplication of the StandardConceptFilter enums between the client and service... at the moment we don't do any swagger-codegen in common-api so this is necessary. At some point we may move the API code into common-api as well, and then we can get rid of this.

File: api/src/main/resources/client_api.yaml [hunk 16]
```diff
@@ -74,6 +74,70 @@ paths:
           schema:
             $ref: "#/definitions/MaterializeCohortResponse"
 
+  /v1/workspaces/{workspaceNamespace}/{workspaceId}/concepts:
+      post:
```
Comment (calbach) [inline api/src/main/resources/client_api.yaml]: Wasn't this a GET previously?

If you're going to make it a POST, then you should use a custom verb at the end of the URL, e.g. `.../{workspaceId}/searchConcepts` or `.../{workspaceId}/concepts:search`. POST to a noun (what you have currently) generally means "create that noun"

File: api/src/main/java/org/pmiops/workbench/api/ConceptsController.java [hunk 17]
```diff
@@ -20,6 +23,26 @@
 @RestController
 public class ConceptsController implements ConceptsApiDelegate {
 
+  private static final ImmutableMultimap<Domain, String> DOMAIN_MAP =
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/api/ConceptsController.java]: We may want to in future have this kind of mapping in our CDR schema somewhere... I'm not sure. For now, hardcoding it here seems OK-ish.

My assumption is that researchers are going to want to filter concepts by the enum values shown here, and not have to choose from underlying domain IDs (which in a number of cases could be for two different domains.)

File: api/src/main/resources/client_api.yaml [hunk 18]
```diff
@@ -74,6 +74,70 @@ paths:
...
+            name: workspaceNamespace
+            description: The namespace of the workspace for which concepts are being looked up.
+            required: true
+            type: string
+          - in: path
+            name: workspaceId
+            description: The ID of the workspace for which concepts are being looked up.
+            required: true
+            type: string
+          - in: body
```
Comment (calbach) [inline api/src/main/resources/client_api.yaml]: If you're going with POST, please make this into a JSON body with a request model definition as we have elsewhere. For consistency, but also this makes the swagger codegen better (Java controller method currently has N parameters).

File: api/src/main/resources/client_api.yaml [hunk 19]
```diff
@@ -74,6 +74,70 @@ paths:
...
+            name: query
+            schema:
+              type: string
+            required: true
+            description: >
+              A query string that can be used to match a subset of the name (case-insensitively),
+              the entire code value (case-insensitively), or the concept ID.
+          - in: body
+            name: standard_concept_filter
```
Comment (calbach) [inline api/src/main/resources/client_api.yaml]: Field names should be lower-camel-cased for consistency with the rest of the api

File: api/src/main/resources/client_api.yaml (hunk 16 above)
Comment (danqrodney) [reply api/src/main/resources/client_api.yaml]: Done.

File: api/src/main/resources/client_api.yaml (hunk 18 above)
Comment (danqrodney) [reply api/src/main/resources/client_api.yaml]: Done.

File: api/src/main/resources/client_api.yaml (hunk 19 above)
Comment (danqrodney) [reply api/src/main/resources/client_api.yaml]: Done.
